### PR TITLE
32093 max ent groups

### LINKED ISSUE #32093: Muon MaxEnt with groups
**Is your feature request related to a problem? Please describe.**

In the frequency domain analysis GUI, the user can only select a "raw" option for calculating the frequency spectrum. However, the algorithm does allow for  it to be done via groups and this should be exposed into the GUI.
**Describe the solution you'd like**

We should remove the current workspace selection and replace it with two new rows:

1. The run number
2. Calculate by; with the options "by detectors" or "by groups"

**Describe alternatives you've considered**

We should not assume its for all runs due to the computational cost of maxent. 
**Additional context**

We will need to generate the group table from the grouping tab. However, it will need the full list of detectors i.e. instead of 1-3 it will need 1,2,3 in the table. 

Example script:
```
Load(Filename='MUSR00022725.nxs', OutputWorkspace='MUSR00022725')
# estimate phases
tab = CreateEmptyTableWorkspace()
tab.addColumn('str', 'Detectors')
tab.addRow(["1,2,3,4"])
tab.addRow(["25,32,41,48"])
tab.addRow(["1,8,33,40"])
tab.addRow(["9,16,57,64"])

CalMuonDetectorPhases(InputWorkspace='MUSR00022725', FirstGoodData=0.10000000000000001, LastGoodData=16, DetectorTable='phases', DataFitted='fitted', ForwardSpectra='9-16,57-64', BackwardSpectra='25-32,41-48')
MuonMaxent(InputWorkspace='MUSR00022725', InputPhaseTable='phases', Npts='16384', OuterIterations='9', InnerIterations='12', DefaultLevel=0.11, Factor=1.03, OutputWorkspace='freq', OutputPhaseTable='phasesOut', ReconstructedSpectra='time')
MuonMaxent(InputWorkspace='MUSR00022725', Npts='16384', OuterIterations='9', InnerIterations='12', DefaultLevel=0.11, Factor=1.03, OutputWorkspace='f_grouos', ReconstructedSpectra='time_groups', GroupTable=tab)


```

### PR DESCRIPTION
**Pull request by AnthonyLim23**
**Description of work.**

The muon group want to be able to use the groups (e.g. fwd, bwd) for MaxEnt calculations. This PR adds the code needed to allow this. 

This PR also fixes a bug that meant that the phase tables from MuonMaxEnt were not available in phase tab.
**To test:**

Open frequency domain analysis
Load `MUSR 62260`
Go to the transform tab
Change the mode from FFT to MaxEnt
Tick the `Output phase table` option
Click calculate
You should get a nice peak in the plot
In the top table, the phase table row will now have an extra option. Select it
Click calculate and you will get a nicer peak
Go to the grouping tab and unselect/add a group (all that matters is that the number of selected groups changes)
Go back to the transform tab
For the phase table row, there will only be one option "None"
Click calculate 
A new option will appear in the phase table row
Set the number of selected groups back to 4 on the grouping tab
On the transform tab, the phase table row will have the original phase table as an option

Open frequency domain analysis
Load `HIFI 84447`
Go to the transform tab
Change the mode from FFT to MaxEnt
Tick the `Output phase table` option
Click calculate
You should get a nice peak in the plot
In the top table, the phase table row will now have an extra option. Select it
Click calculate and you will get a nicer peak
Use the period selector to change the period
Do another maxent calculation - the result should change

Open frequency domain analysis
Load `HIFI 134028-30`
Go to the transform tab
Change the mode from FFT to MaxEnt
Tick the `Output phase table` option
Click calculate
You should get a nice peak in the plot
Change the workspace selector to the next run
Click calculate - the peak will change

Open frequency domain analysis
Load `MUSR 62260`
Go to the transform tab
Change the mode from FFT to MaxEnt
Tick the `Output phase table` option
Change the 'method' row to All detectors
Click calculate
Check that a phase table has appeared in the phase table row options
Go to the phase tab
You should have a phase table to use for phasequad. 
For the other tests, you will not get a phase table in the phasequad tab (it needs the method to be all detectors)

Fixes #32093. 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/Muon/GUI/FrequencyDomainAnalysis/MaxEnt/maxent_presenter.py [hunk 1]
```diff
@@ -29,39 +41,67 @@ class MaxEntPresenter(object):
...
         self.getWorkspaceNames()
 
     def clear(self):
-        self.view.addItems([])
+        self.view.addRuns([])
         self.view.update_phase_table_combo(['Construct'])
 
     # functions
     def getWorkspaceNames(self):
-        final_options = self.load.getGroupedWorkspaceNames()
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/FrequencyDomainAnalysis/MaxEnt/maxent_presenter.py:60]: check it call load and runs

File: scripts/Muon/GUI/FrequencyDomainAnalysis/MaxEnt/maxent_presenter.py [hunk 2]
```diff
@@ -99,36 +139,69 @@ def handleMaxEntButton(self):
     def handleFinished(self):
         self.activate()
         self.calculation_finished_notifier.notify_subscribers(self._maxent_output_workspace_name)
+        # if phase table is outputed
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/FrequencyDomainAnalysis/MaxEnt/maxent_presenter.py:142]: check ifs

File: scripts/Muon/GUI/FrequencyDomainAnalysis/MaxEnt/maxent_presenter.py [hunk 3]
```diff
@@ -99,36 +139,69 @@ def handleMaxEntButton(self):
...
 
     def calculate_maxent(self, alg):
         maxent_parameters = self.get_parameters_for_maxent_calculation()
-        base_name = get_maxent_workspace_name(maxent_parameters['InputWorkspace'])
-
-        maxent_workspace = run_MuonMaxent(maxent_parameters, alg, base_name)
-
-        self.add_maxent_workspace_to_ADS(maxent_parameters['InputWorkspace'], maxent_workspace, alg)
+        base_name = get_maxent_workspace_name(maxent_parameters['InputWorkspace'],  self.view.get_method)
+        if self.use_groups and self.get_num_groups ==0:
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/FrequencyDomainAnalysis/MaxEnt/maxent_presenter.py:162]: check ifs

File: scripts/Muon/GUI/FrequencyDomainAnalysis/MaxEnt/maxent_presenter.py [hunk 4]
```diff
@@ -99,36 +139,69 @@ def handleMaxEntButton(self):
...
+            # this is caught as part of the calculation thread
+            raise ValueError("Please select groups in the grouping tab")
+        else:
+            maxent_workspace = run_MuonMaxent(maxent_parameters, alg, base_name)
+            self.add_maxent_workspace_to_ADS(maxent_parameters['InputWorkspace'], maxent_workspace, alg)
+
+    def _create_group_table(self):
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/FrequencyDomainAnalysis/MaxEnt/maxent_presenter.py:169]: test this

File: scripts/Muon/GUI/FrequencyDomainAnalysis/MaxEnt/maxent_presenter.py [hunk 5]
```diff
@@ -99,36 +139,69 @@ def handleMaxEntButton(self):
...
+        return tab
 
     def get_parameters_for_maxent_calculation(self):
         inputs = {}
-
-        inputs['InputWorkspace'] = self.view.input_workspace
-        run = float(re.search('[0-9]+', inputs['InputWorkspace']).group())
-
-        if self.view.phase_table != 'Construct':
+        run = self.view.get_run
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/FrequencyDomainAnalysis/MaxEnt/maxent_presenter.py:182]: check gives correct args

File: scripts/Muon/GUI/FrequencyDomainAnalysis/MaxEnt/maxent_presenter.py [hunk 6]
```diff
@@ -149,21 +222,29 @@ def get_parameters_for_maxent_calculation(self):
         return inputs
 
     def update_phase_table_options(self):
-        phase_table_list = self.load.phase_context.get_phase_table_list(self.load.data_context.instrument)
-        phase_table_list.insert(0, 'Construct')
+        phase_table_list = []
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/FrequencyDomainAnalysis/MaxEnt/maxent_presenter.py:225]: check inc/ex groups correctly

Comment (AnthonyLim23) [inline scripts/Muon/GUI/FrequencyDomainAnalysis/MaxEnt/maxent_presenter.py:225]: check if

**Review by Pasarus — DISMISSED** [review]
Whilst testing this PR I fail at the line:

`For the phase table row, there will only be one option "None"` , where MUSR6226_raw_data FD; etc is present when it shouldn't be.
After attempting to continue in case it was an issue, I get the following error:
image

**Comment by AnthonyLim23** [comment]
Sorry, I dont understand the steps you have taken to get there.

If you dont select the "output phase table" you will only ever be able to do the "None" option

**Comment by AnthonyLim23** [comment]
unrelated failure

File: scripts/Muon/GUI/FrequencyDomainAnalysis/MaxEnt/maxent_presenter.py [hunk 7]
```diff
@@ -99,36 +139,70 @@ def handleMaxEntButton(self):
...
+            raise ValueError("Please select groups in the grouping tab")
+        else:
+            maxent_workspace = run_MuonMaxent(maxent_parameters, alg, base_name)
+            self.add_maxent_workspace_to_ADS(maxent_parameters['InputWorkspace'], maxent_workspace, alg)
+
+    def _create_group_table(self):
+        tab = create_empty_table(GROUPINGTABLE)
+        tab.addColumn('str', 'Detectors')
+        groups = self.get_selected_groups
+        print("moo", groups)
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/FrequencyDomainAnalysis/MaxEnt/maxent_presenter.py:173]:
```suggestion
```

**Review by Matt-Cumber — CHANGES_REQUESTED** [review]
Co-add doesn't play nice.
Load HIFI84447-8 and tick co add
Go to transform tab
Change FFT to maxent (FFT works with co-add might be worth a look)
Click calculate
Crash (but can continue)

**Review by Matt-Cumber — APPROVED** [review]
Looks good. Test failure is unrelated

**Comment by DannyHindson** [comment]
RHEL7 failure was due to unreliable test so have merged this without waiting for that to pass

**Pull request merged**